src/main/drivers/max7456.c

```diff
@@ -0,0 +1,188 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <stdint.h>
+
+#include "common/printf.h"
+#include "platform.h"
+#include "version.h"
+
+#ifdef USE_MAX7456
+
+#include "drivers/bus_spi.h"
+#include "drivers/system.h"
+
+#include "max7456.h"
+
+#define DISABLE_MAX7456       GPIO_SetBits(MAX7456_CS_GPIO,   MAX7456_CS_PIN)
+#define ENABLE_MAX7456        GPIO_ResetBits(MAX7456_CS_GPIO, MAX7456_CS_PIN)
+
+
+uint16_t max_screen_size;
+uint8_t  video_signal_type = 0;
+uint8_t  max7456_lock = 0;
+char screen[480];
+
+
+uint8_t max7456_send(uint8_t add, uint8_t data) {
+    spiTransferByte(MAX7456_SPI_INSTANCE, add);
+    return spiTransferByte(MAX7456_SPI_INSTANCE, data);
+}
+
+
+void max7456_init(uint8_t system) {
+    uint8_t max_screen_rows;
+    uint8_t srdata = 0;
+    uint16_t x;
+    char buf[30];
+
+    //Minimum spi clock period for max7456 is 100ns (10Mhz)
+    spiSetDivisor(MAX7456_SPI_INSTANCE, SPI_9MHZ_CLOCK_DIVIDER);
+
+    delay(1000);
+    // force soft reset on Max7456
+    ENABLE_MAX7456;
+    max7456_send(VM0_REG, MAX7456_RESET);
+    delay(100);
+
+    srdata = max7456_send(0xA0, 0xFF);
+    if ((0x01 & srdata) == 0x01){     //PAL
+          video_signal_type = VIDEO_MODE_PAL;
+    }
+    else if((0x02 & srdata) == 0x02){ //NTSC
+        video_signal_type = VIDEO_MODE_NTSC;
+    }
+
+    // Override detected type: 0-AUTO, 1-PAL, 2-NTSC
+    switch(system) {
+        case 1:
+            video_signal_type = VIDEO_MODE_PAL;
+            break;
+        case 2:
+            video_signal_type = VIDEO_MODE_NTSC;
+            break;
+    }
+
+    if (video_signal_type) { //PAL
+        max_screen_size = 480;
+        max_screen_rows = 16;
+    } else {                 // NTSC
+        max_screen_size = 390;
+        max_screen_rows = 13;
+    }
+
+    // set all rows to same charactor black/white level
+    for(x = 0; x < max_screen_rows; x++) {
+        max7456_send(MAX7456ADD_RB0 + x, BWBRIGHTNESS);
+    }
+
+    // make sure the Max7456 is enabled
+    max7456_send(VM0_REG, OSD_ENABLE | video_signal_type);
+
+    DISABLE_MAX7456;
+    delay(100);
+
+    x =  160;
+    for (int i = 1; i < 5; i++) {
+        for (int j = 3; j < 27; j++)
+            screen[i * 30 + j] = (char)x++;
+    }
+    tfp_sprintf(buf, "BF VERSION: %s", FC_VERSION_STRING);
+    max7456_write_string(buf, 5*30+5);
+    max7456_write_string("MENU: THRT MID", 6*30+7);
+    max7456_write_string("YAW RIGHT", 7*30+13);
+    max7456_write_string("PITCH UP", 8*30+13);
+    max7456_draw_screen();
+}
+
+// Copy string from ram into screen buffer
+void max7456_write_string(const char *string, int16_t address) {
+    char *dest;
+
+    if (address >= 0)
+        dest  = screen + address;
+    else
+        dest  = screen + (max_screen_size + address);
```

Comment:
negative address used for printing from bottom. F.e. You want to print some value at the very bottom line - you set address to -30. Useful when you arrange screen values unrelated to PAL or NTSC (different number of lines)